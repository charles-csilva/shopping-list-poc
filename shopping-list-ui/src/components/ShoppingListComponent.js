import Button from 'react-bootstrap/Button';
import { useDispatch } from 'react-redux';
import {
	activeItemDeleteAction,
	activeItemMarkAsCompleteAction,
} from '../redux/activeItemsSlice';

export const ShoppingListComponent = ({ data, isCompleteAction }) => {
	const dispatch = useDispatch();
	return (
		<div className="d-flex justify-content-center">
			<table className="table table-bordered">
				<tbody>
					{data.length === 0 && <p>(No Items)</p>}
					{data.map(({ id, name, price }) => (
						<tr key={id}>
							<td>
								{`${name} - $${price}`}
								<Button
									className="btn-action"
									variant="link"
									onClick={() =>
										isCompleteAction
											? dispatch(activeItemMarkAsCompleteAction(id))
											: dispatch(activeItemDeleteAction(id))
									}
								>
									{isCompleteAction && <span>✅</span>}
									{!isCompleteAction && <span>❌</span>}
								</Button>
							</td>
						</tr>
					))}
				</tbody>
			</table>
		</div>
	);
};
