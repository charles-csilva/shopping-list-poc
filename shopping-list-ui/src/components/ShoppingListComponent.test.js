import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';
import { ShoppingListComponent } from './ShoppingListComponent';
import {
	activeItemMarkAsCompleteAction,
	activeItemDeleteAction,
} from '../redux/activeItemsSlice';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => ({
	useSelector: jest.fn(),
	useDispatch: () => mockDispatch,
}));
describe('ShoppingListComponent', () => {
	const mockData = [{ id: 1, name: 'p1', price: 1.23 }];

	describe('must be properly rendered when', () => {
		it('data is an empty array', () => {
			const component = renderer.create(
				<ShoppingListComponent
					data={[]}
					isCompleteAction
				></ShoppingListComponent>
			);
			expect(component).toMatchSnapshot();
		});

		it('data has values and action available is Mark as Completed', () => {
			const fixture = renderer.create(
				<ShoppingListComponent
					data={mockData}
					isCompleteAction
				></ShoppingListComponent>
			);
			expect(fixture).toMatchSnapshot();
		});

		it('data has values and action available is Delete', () => {
			const component = renderer.create(
				<ShoppingListComponent
					data={mockData}
					isCompleteAction={false}
				></ShoppingListComponent>
			);
			expect(component).toMatchSnapshot();
		});
	});

	describe('must handle clicks properly when', () => {
		let container = null;
		beforeEach(() => {
			container = document.createElement('div');
			document.body.appendChild(container);
		});

		afterEach(() => {
			unmountComponentAtNode(container);
			container.remove();
			container = null;
		});
		it('action is Mark as Complete', () => {
			act(() => {
				render(
					<ShoppingListComponent
						data={mockData}
						isCompleteAction
					></ShoppingListComponent>,
					container
				);
			});
			const button = document.querySelector('button.btn-action');
			button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
			expect(mockDispatch).toHaveBeenCalledTimes(1);
			expect(mockDispatch).toHaveBeenCalledWith(
				activeItemMarkAsCompleteAction(mockData[0].id)
			);
		});

		it('action is Delete', () => {
			act(() => {
				render(
					<ShoppingListComponent
						data={mockData}
						isCompleteAction={false}
					></ShoppingListComponent>,
					container
				);
			});
			const button = document.querySelector('button.btn-action');
			button.dispatchEvent(new MouseEvent('click', { bubbles: true }));
			expect(mockDispatch).toHaveBeenCalledTimes(1);
			expect(mockDispatch).toHaveBeenCalledWith(
				activeItemDeleteAction(mockData[0].id)
			);
		});
	});
});
