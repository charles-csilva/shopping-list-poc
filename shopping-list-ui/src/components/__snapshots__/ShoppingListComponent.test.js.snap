// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingListComponent must be properly rendered when data has values and action available is Delete 1`] = `
<div
  className="d-flex justify-content-center"
>
  <table
    className="table table-bordered"
  >
    <tbody>
      <tr>
        <td>
          p1 - $1.23
          <button
            className="btn-action btn btn-link"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <span>
              ❌
            </span>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ShoppingListComponent must be properly rendered when data has values and action available is Mark as Completed 1`] = `
<div
  className="d-flex justify-content-center"
>
  <table
    className="table table-bordered"
  >
    <tbody>
      <tr>
        <td>
          p1 - $1.23
          <button
            className="btn-action btn btn-link"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <span>
              ✅
            </span>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ShoppingListComponent must be properly rendered when data is an empty array 1`] = `
<div
  className="d-flex justify-content-center"
>
  <table
    className="table table-bordered"
  >
    <tbody>
      <p>
        (No Items)
      </p>
    </tbody>
  </table>
</div>
`;
